test/common/http/dispatcher_test.cc

```diff
@@ -739,12 +850,189 @@ TEST_F(DispatcherTest, MultipleStreams) {
+  ASSERT_EQ(cc.on_headers_calls, 0);
+  stream_deletion_post_cb();
+
+  // Ensure that the callbacks on the bridge_callbacks were called.
+  ASSERT_EQ(cc.on_complete_calls, 0);
+  ASSERT_EQ(cc.on_error_calls, 1);
+}
+
+TEST_F(DispatcherTest, EnvoyLocalReplyWithData) {
+  ready();
+
+  envoy_stream_t stream = 1;
+  // Setup bridge_callbacks to handle the response headers.
+  envoy_http_callbacks bridge_callbacks;
+  callbacks_called cc = {0, 0, 0, 0, 0, 0};
+  bridge_callbacks.context = &cc;
+  bridge_callbacks.on_error = [](envoy_error error, void* context) -> void {
+    ASSERT_EQ(error.error_code, ENVOY_CONNECTION_FAILURE);
+    ASSERT_EQ(Http::Utility::convertToString(error.message), "error message");
+    callbacks_called* cc = static_cast<callbacks_called*>(context);
+    cc->on_error_calls++;
+    error.message.release(error.message.context);
+  };
+
+  // Build a set of request headers.
+  TestRequestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  envoy_headers c_headers = Utility::toBridgeHeaders(headers);
+
+  // Create a stream.
+  Event::PostCb start_stream_post_cb;
+  EXPECT_CALL(event_dispatcher_, post(_)).WillOnce(SaveArg<0>(&start_stream_post_cb));
+  EXPECT_EQ(http_dispatcher_.startStream(stream, bridge_callbacks), ENVOY_SUCCESS);
+
+  // Grab the response encoder in order to dispatch responses on the stream.
+  // Return the request decoder to make sure calls are dispatched to the decoder via the dispatcher
+  // API.
+  EXPECT_CALL(api_listener_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder& encoder, bool) -> RequestDecoder& {
+        response_encoder_ = &encoder;
+        return request_decoder_;
+      }));
+  start_stream_post_cb();
+
+  // Send request headers.
+  Event::PostCb send_headers_post_cb;
+  EXPECT_CALL(event_dispatcher_, post(_)).WillOnce(SaveArg<0>(&send_headers_post_cb));
+  http_dispatcher_.sendHeaders(stream, c_headers, true);
+
+  EXPECT_CALL(request_decoder_, decodeHeaders_(_, true));
+  send_headers_post_cb();
+
+  // Encode response headers. A non-200 code triggers an on_error callback chain. In particular, a
+  // 503 should have an ENVOY_CONNECTION_FAILURE error code. However, do not end the stream yet.
+  TestResponseHeaderMapImpl response_headers{{":status", "503"}};
+  response_encoder_->encodeHeaders(response_headers, false);
+  ASSERT_EQ(cc.on_headers_calls, 0);
+
+  Event::PostCb stream_deletion_post_cb;
+  EXPECT_CALL(event_dispatcher_, isThreadSafe()).Times(1).WillRepeatedly(Return(true));
+  EXPECT_CALL(event_dispatcher_, post(_)).WillOnce(SaveArg<0>(&stream_deletion_post_cb));
+  Buffer::InstancePtr response_data{new Buffer::OwnedImpl("error message")};
+  response_encoder_->encodeData(*response_data, true);
+  ASSERT_EQ(cc.on_data_calls, 0);
+  stream_deletion_post_cb();
+
+  // Ensure that the callbacks on the bridge_callbacks were called.
+  ASSERT_EQ(cc.on_complete_calls, 0);
+  ASSERT_EQ(cc.on_error_calls, 1);
```

Comment:
I agree with @buildbreaker's comment above regarding encapsulating testable pieces so we don't have large test functions. Another idea might be to break up this file with those smaller individual tests (it's ~1500LOC)

Reply:
see https://github.com/lyft/envoy-mobile/pull/792#discussion_r405148553